Source/santad/EventProviders/SNTEndpointSecurityManager.mm

```diff
@@ -127,94 +149,103 @@ - (void)messageHandler:(const es_message_t *)m API_AVAILABLE(macos(10.15)) {
+      p = [p stringByAppendingPathComponent:@(m->event.link.target_filename.data)];
+      [self populateBufferFromString:p.UTF8String
+                              length:p.length
+                              buffer:sm.newpath
+                                size:sizeof(sm.newpath)];
       callback = self.logCallback;
-      audit_token = m->process->audit_token;
       break;
     }
     case ES_EVENT_TYPE_NOTIFY_RENAME: {
       sm.action = ACTION_NOTIFY_RENAME;
-      sm.ppid = m->process->original_ppid;
-      strncpy(sm.path, m->event.rename.source->path.data, m->event.rename.source->path.length);
-      sm.path[m->event.rename.source->path.length] = '\0';
+      targetFile = m->event.rename.source;
+      targetProcess = m->process;
 
       switch(m->event.rename.destination_type) {
-        case ES_DESTINATION_TYPE_NEW_PATH:
-          strncpy(sm.newpath, m->event.rename.destination.new_path.filename.data, m->event.rename.destination.new_path.filename.length);
-          sm.newpath[m->event.rename.destination.new_path.filename.length] = '\0';
+        case ES_DESTINATION_TYPE_NEW_PATH: {
+          NSString *p = @(m->event.rename.destination.new_path.dir->path.data);
+          p = [p stringByAppendingPathComponent:
+                  @(m->event.rename.destination.new_path.filename.data)];
+          [self populateBufferFromString:p.UTF8String
+                                  length:p.length
+                                  buffer:sm.newpath
+                                    size:sizeof(sm.newpath)];
           break;
-        case ES_DESTINATION_TYPE_EXISTING_FILE:
-          strncpy(sm.newpath, m->event.rename.destination.existing_file->path.data, m->event.rename.destination.existing_file->path.length);
-          sm.newpath[m->event.rename.destination.existing_file->path.length] = '\0';
+        }
+        case ES_DESTINATION_TYPE_EXISTING_FILE: {
+          [self populateBufferFromESFile:m->event.rename.destination.existing_file
+                                  buffer:sm.newpath
+                                    size:sizeof(sm.newpath)];
           break;
+        }
       }
-
       callback = self.logCallback;
-      audit_token = m->process->audit_token;
       break;
     }
     default:
       break;
   }
 
-  // Filter events matching the prefix tree.
-  // Exec events are not filtered.
+  // Deny auth exec events if the path doesn't fit in the santa message.
+  // TODO(bur/rah): Add support for larger paths.
+  if ([self populateBufferFromESFile:targetFile buffer:sm.path size:sizeof(sm.path)]
+      && m->event_type == ES_EVENT_TYPE_AUTH_EXEC) {
+    LOGE(@"path is truncated, deny: %s", sm.path);
+    es_respond_auth_result(self.client, m, ES_AUTH_RESULT_DENY, false);
+    es_free_message(m);
+    return;
+  }
+
+  // Filter file op events matching the prefix tree.
   if (!(m->event_type == ES_EVENT_TYPE_AUTH_EXEC || m->event_type == ES_EVENT_TYPE_NOTIFY_EXEC) &&
```

Comment:
How come the prefix tree is applying to auth events?

Reply:
It doesn't apply to any exec event, note the negation of the (||) clause.

Reply:
Ah I missed that